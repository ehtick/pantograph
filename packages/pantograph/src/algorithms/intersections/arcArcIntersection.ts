import { Vector } from "../../definitions.js";
import { Arc } from "../../models/segments/Arc.js";
import removeDuplicatePoints from "../../utils/removeDuplicatePoints.js";
import {
  add,
  distance,
  normalize,
  perpendicular,
  sameVector,
  scalarMultiply,
  subtract,
} from "../../vectorOperations";

const complementArc = (arc: Arc): Arc => {
  const { firstPoint, lastPoint, center, clockwise } = arc;
  return new Arc(lastPoint, firstPoint, center, clockwise, {
    ignoreChecks: true,
  });
};

const handleOverlaps = (arc1: Arc, arc2: Arc): Arc[] => {
  // handle the case with common points at start or end first
  if (arc1.isSame(arc2)) {
    return [arc1];
  }
  // two arcs with common points at start and end (but not the same arc)

  const points = removeDuplicatePoints(
    [
      arc2.isOnSegment(arc1.firstPoint) ? arc1.firstPoint : null,
      arc2.isOnSegment(arc1.lastPoint) ? arc1.lastPoint : null,
      arc1.isOnSegment(arc2.firstPoint) ? arc2.firstPoint : null,
      arc1.isOnSegment(arc2.lastPoint) ? arc2.lastPoint : null,
    ].filter((p) => p !== null) as Vector[],
    // We sort by the param value of the first arc. This means that the points
    // will be sorted with the same orientation than arc1
  ).sort((a, b) => arc1.pointToParam(a) - arc1.pointToParam(b));

  if (points.length === 0) return [];
  // We consider the case when the arcs touch only on
  // the last point. We consider that they do not overlap there
  //
  // We might want to revisit this choice
  else if (points.length === 1) return [];
  else if (points.length === 2) {
    // Similar to the case with length 1, we ignore the double overlapping point
    if (arc1.isSame(complementArc(arc2))) return [];
    return [new Arc(points[0], points[1], arc1.center, arc1.clockwise)];
  } else if (points.length === 3) {
    // Similar to the case with length 1, we ignore the single overlapping point
    const startIndex =
      sameVector(points[0], arc2.lastPoint) ||
      sameVector(points[0], arc2.firstPoint)
        ? 1
        : 0;
    return [
      new Arc(
        points[0 + startIndex],
        points[1 + startIndex],
        arc1.center,
        arc1.clockwise,
      ),
    ];
  } else if (points.length === 4) {
    return [
      new Arc(points[0], points[1], arc1.center, arc1.clockwise),
      new Arc(points[2], points[3], arc1.center, arc1.clockwise),
    ];
  }
  throw new Error("Bug in the arc arc overlap algorithm");
};

export function arcArcIntersection(
  arc1: Arc,
  arc2: Arc,
  includeOverlaps = false,
  precision?: number,
): Vector[] | Arc[] {
  const epsilon = precision ? precision : arc1.precision;
  const centersDistance = distance(arc1.center, arc2.center);

  const radiusSum = arc1.radius + arc2.radius;

  // The circles do not touch
  if (centersDistance > radiusSum + epsilon) {
    return [];
  }

  const radiusDifference = Math.abs(arc1.radius - arc2.radius);

  // The arcs are concentric
  if (centersDistance < radiusDifference - epsilon) {
    return [];
  }

  // With a common center we can have overlaps
  if (centersDistance < epsilon) {
    if (radiusDifference > epsilon) {
      return [];
    } else {
      if (!includeOverlaps) {
        return [];
      }
      return handleOverlaps(arc1, arc2);
    }
  }

  const centersVector = normalize(subtract(arc2.center, arc1.center));
  // The circles are tangent to each other
  const isOutsideTangent = centersDistance > radiusSum - epsilon;
  if (
    // circles are outside each other
    isOutsideTangent ||
    // circles are inside each other
    Math.abs(centersDistance - radiusDifference) < epsilon
  ) {
    const orientation = isOutsideTangent || arc1.radius > arc2.radius ? 1 : -1;
    const intersectionPoint = add(
      arc1.center,
      scalarMultiply(centersVector, orientation * arc1.radius),
    );

    if (
      arc1.isOnSegment(intersectionPoint) &&
      arc2.isOnSegment(intersectionPoint)
    ) {
      return [intersectionPoint];
    } else {
      return [];
    }
  }

  // The circles cross each other
  const radiusToChord =
    (arc1.radius * arc1.radius) / (2 * centersDistance) -
    (arc2.radius * arc2.radius) / (2 * centersDistance) +
    centersDistance / 2;

  const midPoint = add(
    arc1.center,
    scalarMultiply(centersVector, radiusToChord),
  );

  const halfChord = Math.sqrt(
    arc1.radius * arc1.radius - radiusToChord * radiusToChord,
  );

  const chordVector = perpendicular(centersVector);

  const p1 = add(midPoint, scalarMultiply(chordVector, halfChord));
  const p2 = add(midPoint, scalarMultiply(chordVector, -halfChord));

  const intersections = [];
  if (arc1.isOnSegment(p1) && arc2.isOnSegment(p1)) {
    intersections.push(p1);
  }
  if (arc1.isOnSegment(p2) && arc2.isOnSegment(p2)) {
    intersections.push(p2);
  }

  return intersections;
}
