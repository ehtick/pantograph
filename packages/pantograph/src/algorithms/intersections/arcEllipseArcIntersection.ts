import type { EllipseArc } from "../../models/segments/EllipseArc";
import type { Arc } from "../../models/segments/Arc";
import { ellipseEllipseIntersection } from "./ellipseEllipseIntersection";

export function arcEllipseArcIntersection(arc: Arc, ellipseArc: EllipseArc) {
  const points = ellipseEllipseIntersection(arc, ellipseArc);
  return points.filter((p) => arc.isOnSegment(p) && ellipseArc.isOnSegment(p));
}
