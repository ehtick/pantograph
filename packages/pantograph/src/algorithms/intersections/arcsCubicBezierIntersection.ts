import { EllipseArc } from "../../models/segments/EllipseArc.js";
import { Arc } from "../../models/segments/Arc.js";
import { Vector } from "../../definitions.js";
import removeDuplicateValues from "../../utils/removeDuplicateValues.js";
import { solveGenericPolynomial } from "../solvers/solvePolynomials.js";
import { CubicBezier } from "../../models/segments/CubicBezier.js";

/*
 * We use wolfram alpha to do the analitical resolution of the intersection
 * Collect[ Eliminate[{a*x^2+b*x*y+c*y^2+d * x + e * y + f == 0, p3 * t^3 + p2 * t^2  + p1 * t +p0 == x,q3*t^3+q2*t^2+q1*t+q0==y}, {x,y}], t ]
 * */

const intersectionsPolynomial = (arc: EllipseArc | Arc, curve: CubicBezier) => {
  const [[p0, p1, p2, p3], [q0, q1, q2, q3]] = curve.polynomialCoefficients;

  const el1 = arc.coefficients;
  const a = el1.x2;
  const b = el1.xy;
  const c = el1.y2;
  const d = el1.x;
  const e = el1.y;
  const f = el1.c;

  const p02 = p0 * p0;
  const p12 = p1 * p1;
  const p22 = p2 * p2;
  const p32 = p3 * p3;
  const q02 = q0 * q0;
  const q12 = q1 * q1;
  const q22 = q2 * q2;
  const q32 = q3 * q3;

  const z0 = f + d * p0 + a * p02 + e * q0 + b * p0 * q0 + c * q02;
  const z1 =
    d * p1 +
    2 * a * p0 * p1 +
    b * p1 * q0 +
    e * q1 +
    b * p0 * q1 +
    2 * c * q0 * q1;
  const z2 =
    a * p12 +
    d * p2 +
    2 * a * p0 * p2 +
    b * p2 * q0 +
    b * p1 * q1 +
    c * q12 +
    e * q2 +
    b * p0 * q2 +
    2 * c * q0 * q2;
  const z3 =
    2 * a * p1 * p2 +
    d * p3 +
    2 * a * p0 * p3 +
    b * p3 * q0 +
    b * p2 * q1 +
    b * p1 * q2 +
    2 * c * q1 * q2 +
    e * q3 +
    b * p0 * q3 +
    2 * c * q0 * q3;
  const z4 =
    a * p22 +
    2 * a * p1 * p3 +
    b * p3 * q1 +
    b * p2 * q2 +
    c * q22 +
    b * p1 * q3 +
    2 * c * q1 * q3;
  const z5 = 2 * a * p2 * p3 + b * p3 * q2 + b * p2 * q3 + 2 * c * q2 * q3;
  const z6 = a * p32 + b * p3 * q3 + c * q32;

  return [z0, z1, z2, z3, z4, z5, z6];
};

export function arcsCubicBezierIntersection(
  arc: EllipseArc | Arc,
  curve: CubicBezier
): Vector[] {
  const epsilon = Math.max(arc.precision, curve.precision);

  const polynomial = intersectionsPolynomial(arc, curve);
  const solutions = solveGenericPolynomial(polynomial, epsilon).filter((t) => {
    return t >= -curve.precision && t <= 1 + curve.precision;
  });

  return removeDuplicateValues(solutions, epsilon)
    .map((t) => {
      return curve.paramPoint(t);
    })
    .filter((p) => {
      return arc.isOnSegment(p);
    });
}
