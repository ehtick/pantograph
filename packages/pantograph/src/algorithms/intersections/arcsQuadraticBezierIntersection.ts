import { EllipseArc } from "../../models/segments/EllipseArc.js";
import { Arc } from "../../models/segments/Arc.js";
import { Vector } from "../../definitions.js";
import removeDuplicateValues from "../../utils/removeDuplicateValues.js";
import { solveQuartic } from "../solvers/solvePolynomials.js";
import { QuadraticBezier } from "../../models/segments/QuadraticBezier.js";

/*
 * We use wolfram alpha to do the analitical resolution of the intersection
 * Collect[ Eliminate[{a*x^2+b*x*y+c*y^2+d * x + e * y + f == 0, p2 * t^2  + p1 * t +p0 == x,q2*t^2+q1*t+q0==y}, {x,y}], t ]
 * */

const intersectionsPolynomial = (
  arc: EllipseArc | Arc,
  curve: QuadraticBezier
): [number, number, number, number, number] => {
  const [[p0, p1, p2], [q0, q1, q2]] = curve.polynomialCoefficients;

  const el1 = arc.coefficients;
  const a = el1.x2;
  const b = el1.xy;
  const c = el1.y2;
  const d = el1.x;
  const e = el1.y;
  const f = el1.c;

  const p02 = p0 * p0;
  const p12 = p1 * p1;
  const p22 = p2 * p2;
  const q02 = q0 * q0;
  const q12 = q1 * q1;
  const q22 = q2 * q2;

  const z0 = a * p02 + b * p0 * q0 + c * q02 + d * p0 + e * q0 + f;
  const z1 =
    2 * a * p0 * p1 +
    b * p0 * q1 +
    b * p1 * q0 +
    2 * c * q0 * q1 +
    d * p1 +
    e * q1;
  const z2 =
    2 * a * p0 * p2 +
    a * p12 +
    b * p0 * q2 +
    b * p1 * q1 +
    b * p2 * q0 +
    2 * c * q0 * q2 +
    c * q12 +
    d * p2 +
    e * q2;

  const z3 = 2 * a * p1 * p2 + b * p1 * q2 + b * p2 * q1 + 2 * c * q1 * q2;
  const z4 = a * p22 + b * p2 * q2 + c * q22;

  return [z0, z1, z2, z3, z4];
};

export function arcsQuadraticBezierIntersection(
  arc: EllipseArc | Arc,
  curve: QuadraticBezier
): Vector[] {
  const epsilon = Math.max(arc.precision, curve.precision);

  const polynomial = intersectionsPolynomial(arc, curve);
  const solutions = solveQuartic(...polynomial).filter((t) => {
    return t >= -curve.precision && t <= 1 + curve.precision;
  });

  return removeDuplicateValues(solutions, epsilon)
    .map((t) => {
      return curve.paramPoint(t);
    })
    .filter((p) => {
      return arc.isOnSegment(p);
    });
}
