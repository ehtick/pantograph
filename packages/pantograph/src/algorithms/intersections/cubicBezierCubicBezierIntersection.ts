import { Vector } from "../../definitions.js";
import { CubicBezier } from "../../models/segments/CubicBezier.js";
import { sameVector } from "../../vectorOperations.js";
import { bezierClip } from "./bezierClip.js";

export function handleOverlaps(curve1: CubicBezier, curve2: CubicBezier) {
  const commonPoints: Vector[] = [];
  const toTest: [Vector, CubicBezier][] = [
    [curve1.firstPoint, curve2],
    [curve1.lastPoint, curve2],
    [curve2.firstPoint, curve1],
    [curve2.lastPoint, curve1],
  ];

  // This could be optimised by doing only one vector => parameter conversion

  toTest.forEach(([point, curve]) => {
    if (curve.isOnSegment(point)) {
      commonPoints.push(point);
    }
  });

  if (commonPoints.length < 2) {
    // no overlap
    return null;
  }

  if (commonPoints.length === 2) {
    return [curve1.splitAt(commonPoints)[1]];
  }

  if (commonPoints.length === 3) {
    // there is one curve that is within the other
    if (
      sameVector(commonPoints[0], curve1.firstPoint) &&
      sameVector(commonPoints[1], curve1.lastPoint)
    ) {
      return [curve1];
    }
    return [curve2];
  }

  if (commonPoints.length === 4) {
    return [curve1];
  }
}

export function cubicBezierCubicBezierIntersection(
  curve1: CubicBezier,
  curve2: CubicBezier,
  includeOverlaps = false
) {
  const epsilon = Math.max(curve1.precision, curve2.precision);

  if (includeOverlaps) {
    const overlappingCurve = handleOverlaps(curve1, curve2);
    if (overlappingCurve) {
      return overlappingCurve;
    }
  }

  return bezierClip(curve1, curve2, epsilon);
}
