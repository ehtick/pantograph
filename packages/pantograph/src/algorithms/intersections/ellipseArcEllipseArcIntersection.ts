import { Vector } from "../../definitions.js";
import { EllipseArc } from "../../models/segments/EllipseArc.js";
import removeDuplicatePoints from "../../utils/removeDuplicatePoints.js";
import { sameVector } from "../../vectorOperations.js";
import { ellipseEllipseIntersection } from "./ellipseEllipseIntersection.js";

const complementArc = (arc: EllipseArc): EllipseArc => {
  const {
    firstPoint,
    lastPoint,
    center,
    majorRadius,
    minorRadius,
    tiltAngle,
    clockwise,
  } = arc;
  return new EllipseArc(
    lastPoint,
    firstPoint,
    center,
    majorRadius,
    minorRadius,
    tiltAngle,
    clockwise,
    {
      ignoreChecks: true,
      angleUnits: "rad",
    }
  );
};

const handleOverlaps = (arc1: EllipseArc, arc2: EllipseArc): EllipseArc[] => {
  // handle the case with common points at start or end first
  if (arc1.isSame(arc2)) {
    return [arc1];
  }
  // two arcs with common points at start and end (but not the same arc)

  const partialArc = (firstPoint: Vector, lastPoint: Vector) =>
    new EllipseArc(
      firstPoint,
      lastPoint,
      arc1.center,
      arc1.majorRadius,
      arc1.minorRadius,
      arc1.tiltAngle,
      arc1.clockwise,
      { ignoreChecks: true, angleUnits: "rad" }
    );

  const points = removeDuplicatePoints(
    [
      arc2.isOnSegment(arc1.firstPoint) ? arc1.firstPoint : null,
      arc2.isOnSegment(arc1.lastPoint) ? arc1.lastPoint : null,
      arc1.isOnSegment(arc2.firstPoint) ? arc2.firstPoint : null,
      arc1.isOnSegment(arc2.lastPoint) ? arc2.lastPoint : null,
    ].filter((p) => p !== null) as Vector[]
    // We sort by the param value of the first arc. This means that the points
    // will be sorted with the same orientation than arc1
  ).sort((a, b) => arc1.pointToParam(a) - arc1.pointToParam(b));

  if (points.length === 0) return [];
  // We consider the case when the arcs touch only on
  // the last point. We consider that they do not overlap there
  //
  // We might want to revisit this choice
  else if (points.length === 1) return [];
  else if (points.length === 2) {
    // Similar to the case with length 1, we ignore the double overlapping point
    if (arc1.isSame(complementArc(arc2))) return [];
    return [partialArc(points[0], points[1])];
  } else if (points.length === 3) {
    // Similar to the case with length 1, we ignore the single overlapping point
    const startIndex =
      sameVector(points[0], arc2.lastPoint) ||
      sameVector(points[0], arc2.firstPoint)
        ? 1
        : 0;
    return [partialArc(points[0 + startIndex], points[1 + startIndex])];
  } else if (points.length === 4) {
    return [partialArc(points[0], points[1]), partialArc(points[2], points[3])];
  }
  throw new Error("Bug in the ellipse arc ellipse arc overlap algorithm");
};

export function ellipseArcEllipseArcIntersection(
  arc1: EllipseArc,
  arc2: EllipseArc,
  includeOverlaps = false
): Vector[] | EllipseArc[] {
  const epsilon = Math.max(arc1.precision, arc2.precision);
  const sameEllipse =
    sameVector(arc1.center, arc2.center) &&
    Math.abs(arc1.majorRadius - arc2.majorRadius) < epsilon &&
    Math.abs(arc1.minorRadius - arc2.minorRadius) < epsilon &&
    (Math.abs(arc1.tiltAngle - arc2.tiltAngle) < epsilon ||
      Math.abs(Math.abs(arc1.tiltAngle - arc2.tiltAngle) - Math.PI) < epsilon);
  if (sameEllipse) {
    if (includeOverlaps) return handleOverlaps(arc1, arc2);
    else return [];
  }
  const points = ellipseEllipseIntersection(arc1, arc2);
  return points.filter((p) => arc1.isOnSegment(p) && arc2.isOnSegment(p));
}
