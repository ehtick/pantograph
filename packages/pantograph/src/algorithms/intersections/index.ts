import { Line } from "../../models/segments/Line";

import type { Vector } from "../../definitions";
import { Segment } from "../../models/segments/Segment";
import { Arc } from "../../models/segments/Arc";
import { lineArcIntersection } from "./lineArcIntersection";
import { lineLineIntersection } from "./lineLineIntersection";
import { arcArcIntersection } from "./arcArcIntersection";

export function findIntersections(
  segment1: Segment,
  segment2: Segment,
  precision?: number
): Vector[] {
  if (segment1 instanceof Line && segment2 instanceof Line) {
    const intersection = lineLineIntersection(
      segment1,
      segment2,
      false,
      precision
    );
    if (intersection === null) return [];
    return [intersection as Vector];
  }
  if (segment1 instanceof Line && segment2 instanceof Arc) {
    return lineArcIntersection(segment1, segment2, precision);
  }
  if (segment1 instanceof Arc && segment2 instanceof Line) {
    return lineArcIntersection(segment2, segment1, precision);
  }
  if (segment1 instanceof Arc && segment2 instanceof Arc) {
    return arcArcIntersection(segment1, segment2, false, precision) as Vector[];
  }

  throw new Error("Not implemented");
}

export function findIntersectionsAndOverlaps(
  segment1: Segment,
  segment2: Segment,
  precision?: number
): { intersections: Vector[]; overlaps: Segment[]; count: number } {
  // If we have two lines, checks are fast enough to not use bounding boxes
  if (segment1 instanceof Line && segment2 instanceof Line) {
    const intersection = lineLineIntersection(
      segment1,
      segment2,
      true,
      precision
    );
    if (intersection === null)
      return { intersections: [], overlaps: [], count: 0 };
    if (intersection instanceof Line)
      return { intersections: [], overlaps: [intersection], count: 1 };
    return { intersections: [intersection], overlaps: [], count: 1 };
  }

  if (!segment1.boundingBox.overlaps(segment2.boundingBox)) {
    return { intersections: [], overlaps: [], count: 0 };
  }

  if (segment1 instanceof Line && segment2 instanceof Arc) {
    const intersections = lineArcIntersection(segment1, segment2, precision);
    return { intersections, overlaps: [], count: intersections.length };
  }

  if (segment1 instanceof Arc && segment2 instanceof Line) {
    const intersections = lineArcIntersection(segment2, segment1, precision);
    return { intersections, overlaps: [], count: intersections.length };
  }

  if (segment1 instanceof Arc && segment2 instanceof Arc) {
    const intersections = arcArcIntersection(
      segment1,
      segment2,
      true,
      precision
    );
    if (!intersections.length)
      return { intersections: [], overlaps: [], count: 0 };
    if (intersections[0] instanceof Arc)
      return {
        intersections: [],
        overlaps: intersections as Arc[],
        count: intersections.length,
      };
    return {
      intersections: intersections as Vector[],
      overlaps: [],
      count: intersections.length,
    };
  }

  throw new Error("Not implemented");
}
