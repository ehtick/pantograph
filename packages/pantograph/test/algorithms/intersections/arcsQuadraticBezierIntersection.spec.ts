import { describe, it, expect } from "vitest";

import { arcsQuadraticBezierIntersection } from "../../../src/algorithms/intersections/arcsQuadraticBezierIntersection";
import { Arc } from "../../../src/models/segments/Arc";
import { EllipseArc } from "../../../src/models/segments/EllipseArc";

import { QuadraticBezier } from "../../../src/models/segments/QuadraticBezier";

//import { debugImg, dpnt } from "../../debug";

describe("arcsQuadraticBezierIntersection", () => {
  describe("should return the intersection points of an arc and cubic curve", () => {
    it("works for 2 intersection points", () => {
      const arc = new Arc([0, -1.5], [1.5, 0], [0, 0]).rotate(-30);
      const curve = new QuadraticBezier([0, 0], [1, 1], [0, 1.1]).translate(
        0.5,
        -1.8
      );
      const intersections = arcsQuadraticBezierIntersection(arc, curve);

      //debugImg([arc, curve, ...intersections.map((p) => dpnt(p))]);
      expect(intersections.length).toBe(2);
      expect(intersections).toMatchSnapshot();
    });

    it("works for 1 intersection points", () => {
      const arc = new Arc([0, -1.5], [1.5, 0], [0, 0]).rotate(-30);
      const curve = new QuadraticBezier([0, 0], [1, 1], [0, 1.1]).translate(
        0.5,
        -1.5
      );
      const intersections = arcsQuadraticBezierIntersection(arc, curve);

      //debugImg([arc, curve, ...intersections.map((p) => dpnt(p))]);
      expect(intersections.length).toBe(1);
      expect(intersections).toMatchSnapshot();
    });

    it.fails("works for a tangent points", () => {
      // This fails for now, the precision is not good enough. I should either
      // change the precision or the algorithm.
      const curve = new QuadraticBezier([1, 0], [1, 1], [0, 0]);

      const intersectionPoint = curve.paramPoint(0.5);

      const arc = new Arc([0, -1.5], [1.5, 0], [0, 0])
        .rotate(45)
        .translateX(-1.5)
        .translateTo(intersectionPoint);

      const intersections = arcsQuadraticBezierIntersection(arc, curve);

      //debugImg([arc, curve, ...intersections.map((p) => dpnt(p))]);
      expect(intersections.length).toBe(1);
      expect(intersections[0]).toBeVector(intersectionPoint);
    });

    it("works for no intersection points", () => {
      const arc = new Arc([0, -1.5], [1.5, 0], [0, 0]).rotate(-30);
      const curve = new QuadraticBezier([0, 0], [1, 1], [1, 0.1]).translate(
        0.5,
        0
      );
      const intersections = arcsQuadraticBezierIntersection(arc, curve);

      //debugImg([arc, curve, ...intersections.map((p) => dpnt(p))]);
      expect(intersections.length).toBe(0);
    });
  });

  describe("should return the intersection points of an ellipse arc and cubic curve", () => {
    it("works for 2 intersection points", () => {
      const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, true).rotate(
        5
      );
      const curve = new QuadraticBezier([0, 0], [1, 1], [1, 0.1]).translate(
        -0.1,
        0.2
      );
      const intersections = arcsQuadraticBezierIntersection(arc, curve);

      //debugImg([arc, curve, ...intersections.map((p) => dpnt(p))]);
      expect(intersections.length).toBe(2);
      expect(intersections).toMatchSnapshot();
    });

    it("works for 1 intersection points", () => {
      const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, true).rotate(
        -25
      );
      const curve = new QuadraticBezier([0, 0], [1, 1], [1, 0.1]).translate(
        -0.6,
        -0.5
      );
      const intersections = arcsQuadraticBezierIntersection(arc, curve);

      //debugImg([arc, curve, ...intersections.map((p) => dpnt(p))]);
      expect(intersections.length).toBe(1);
      expect(intersections).toMatchSnapshot();
    });

    it("works for no intersection points", () => {
      const arc = new EllipseArc([0, 0], [2, 1], [2, 0], 2, 1, 0, true).rotate(
        5
      );
      const curve = new QuadraticBezier([0, 0], [1, 1], [1, 0.1]).translate(
        1,
        -0.5
      );
      const intersections = arcsQuadraticBezierIntersection(arc, curve);

      //debugImg([arc, curve, ...intersections.map((p) => dpnt(p))]);
      expect(intersections.length).toBe(0);
    });
  });
});
