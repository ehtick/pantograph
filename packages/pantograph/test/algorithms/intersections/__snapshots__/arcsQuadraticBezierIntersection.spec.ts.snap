// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`arcsCubicBezierIntersection > should return the intersection points of an arc and cubic curve > works for 1 intersection points 1`] = `
[
  [
    0.9856923801013043,
    -1.1306681793568911,
  ],
]
`;

exports[`arcsCubicBezierIntersection > should return the intersection points of an arc and cubic curve > works for 2 intersection points 1`] = `
[
  [
    1.2838079110064191,
    -0.7757817010199046,
  ],
  [
    1.000365116046977,
    -1.1177073116859892,
  ],
]
`;

exports[`arcsCubicBezierIntersection > should return the intersection points of an ellipse arc and cubic curve > works for 1 intersection points 1`] = `
[
  [
    0.22520814189493277,
    0.5233615256349221,
  ],
]
`;

exports[`arcsCubicBezierIntersection > should return the intersection points of an ellipse arc and cubic curve > works for 2 intersection points 1`] = `
[
  [
    0.02865713630405221,
    0.22060248734679613,
  ],
  [
    0.397261614379911,
    0.6717676341327191,
  ],
]
`;

exports[`arcsQuadraticBezierIntersection > should return the intersection points of an arc and cubic curve > works for 1 intersection points 1`] = `
[
  [
    0.5016105460102331,
    -1.4136431162536447,
  ],
]
`;

exports[`arcsQuadraticBezierIntersection > should return the intersection points of an arc and cubic curve > works for 2 intersection points 1`] = `
[
  [
    0.5422330332338472,
    -1.3985647420373581,
  ],
  [
    1.2730265816185793,
    -0.7933494327801038,
  ],
]
`;

exports[`arcsQuadraticBezierIntersection > should return the intersection points of an ellipse arc and cubic curve > works for 1 intersection points 1`] = `
[
  [
    0.3888871282830251,
    0.3191386060425175,
  ],
]
`;

exports[`arcsQuadraticBezierIntersection > should return the intersection points of an ellipse arc and cubic curve > works for 2 intersection points 1`] = `
[
  [
    0.027304605340843904,
    0.2166293077088198,
  ],
  [
    0.8741896602243211,
    0.9314675309798026,
  ],
]
`;
